tests/process-template.js

```diff
@@ -0,0 +1,127 @@
+                        template.forEach(parsedTag => {
+                            stream.write(parsedTag);
+                        });
+                        stream.end();
+                    });
+                }, 10);
+                return stream;
+            }
+
+            return Buffer.of('');
+        };
+
+        let index = 0;
+
+        options = {
+            maxAssetLinks: 3,
+            nextIndex: () => ++index,
+            fragmentTag: 'fragment',
+            parseTemplate: parseTemplate(['fragment', 'nested-fragments'], ['script']),
+            pipeDefinition: (name) => Buffer.from(`<pipe id="${name}" />`),
+            pipeAttributes: (attributes) => ({ id: attributes.id }),
+            pipeInstanceName: 'TailorPipe',
+            asyncStream: new AsyncStream(),
+            handleTag: handleNestedTag,
+            requestFragment: requestFragment
+        };
+        resultStream = processTemplate(mockedRequest, options, {});
+    });
+
+    it('returns a stream', () => {
+        assert(resultStream instanceof stream);
+    });
+
+    describe('result stream', () => {
+        it('produces buffers as result', (done) => {
+            resultStream.on('data', chunk => {
+                assert(chunk instanceof Buffer);
+            });
+            resultStream.on('end', () => {
+                done();
+            });
+            resultStream.on('primary:found', () => {
+                options.asyncStream.end();
+            });
+            resultStream.write({ placeholder: 'pipe' });
+            resultStream.write({ name: 'fragment', attributes: { id: 'f3', async: true, src: 'http://fragment/f3' } });
+            resultStream.write({ name: 'fragment', attributes: { id: 'f1', src: 'http://fragment/f1' } });
+            // resultStream.write({ name: 'fragment', attributes: { id: 'f1', src: 'http://fragment/f1' } });
+            resultStream.write({ name: 'fragment', attributes: { id: 'f2', primary: true, src: 'http://fragment/primary' } });
+            resultStream.write({ name: 'nested-fragments' });
+            resultStream.write({ placeholder: 'async' });
+            resultStream.end();
+        });
+
+        it('notifies when its found in the template', (done) => {
+            const onPrimary = sinon.spy();
+
+            resultStream.on('primary:found', onPrimary);
+            resultStream.on('end', () => {
+                assert.equal(onPrimary.callCount, 1);
+                done();
+            });
+            resultStream.resume();
+            options.asyncStream.end();
+            resultStream.write({ placeholder: 'pipe' });
+            resultStream.write({ name: 'fragment', attributes: { id: 'f2', primary: true, src: 'http://fragment/f2' } });
+            resultStream.write({ placeholder: 'async' });
+            resultStream.end();
+        });
+        it('insert pipe definition in the beginning');
```

Comment:
once we finish the tests for process-template we can go ahead and merge it.. rest looks good to me..

Reply:
Also seem like we're not changing the API much here. Mb minor version update would be suitable.